src/hashset.c

```diff
@@ -1317,24 +1319,29 @@ size_t hashsetScan(hashset *t, size_t cursor, hashsetScanFunction fn, void *priv
                 in_probe_sequence |= b->everfull;
             }
 
-            /* Iterate over indices in larger table that are the expansion of
-             * the index pointed to by the cursor in the smaller table. */
-            do {
-                /* Emit elements in table 1. */
-                bucket *b = &t->tables[table1][cursor & mask1];
-                for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
-                    if (b->presence & (1 << pos)) {
-                        void *emit = emit_ref ? &b->elements[pos] : b->elements[pos];
-                        fn(privdata, emit);
+            if ((growing && !ahead_rehash) || (!growing && !behind_rehash)) {
+                /* Iterate over indices in larger table that are the expansion of
+                 * the index pointed to by the cursor in the smaller table. */
+                size_t big_table_cursor = cursor;
+                do {
+                    /* Emit elements in bigger table. */
+                    bucket *b = &t->tables[table_big][big_table_cursor & mask_big];
+                    for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                        if (b->presence & (1 << pos)) {
+                            void *emit = emit_ref ? &b->elements[pos] : b->elements[pos];
+                            fn(privdata, emit);
+                        }
                     }
-                }
-                in_probe_sequence |= b->everfull;
+                    in_probe_sequence |= b->everfull;
 
-                /* Increment the reverse cursor not covered by the smaller mask.*/
-                cursor = nextCursor(cursor, mask1);
+                    /* Increment the reverse cursor not covered by the smaller mask.*/
+                    big_table_cursor = nextCursor(big_table_cursor, mask_big);
+
+                    /* Continue while bits covered by mask difference is non-zero */
+                } while (big_table_cursor & (mask_small ^ mask_big));
+            }
 
-                /* Continue while bits covered by mask difference is non-zero */
-            } while ((cursor & (mask0 ^ mask1)) && cursor != start_cursor);
+            cursor = nextCursor(cursor, mask_small);
```

Comment:
Why increment cursor again? It was done above already.

I guess it's not done if we skip the big table due to already rehashed, so it needs to be in an `else` clause?

```suggestion
        } else {
            cursor = nextCursor(cursor, mask_small);
```

Reply:
The above sub-loop increments big_table_cursor though, not cursor

Reply:
Ah, there is a big cursor. Ok.